File: 1. String Permutations - GFG/1.-string-permutations.cpp
// { Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution{
    public:
    //Complete this function
    
    /*vector<string> permutation(string S)
    {
        vector<string> res;
        sort(S.begin(), S.end());
        res.push_back(S);
        
        while(next_permutation(S.begin(), S.end()))
          res.push_back(S);
          
        return res;  
        
    }  */
    
    void solve(string S, int s, int e, vector<string>& res)
    {
        if(s==e)
        {
            res.push_back(S);
            return;
        }
        
        for(int i=s; i<=e; i++)
        {
            swap(S[s], S[i]);
            solve(S, s+1, e, res);
            swap(S[s], S[i]);
        }
        
        return;
    }
    
    vector<string> permutation(string S)
    {
        vector<string> res;
        //sort(S.begin(), S.end());
        int n=S.size();
        
        solve(S, 0, n-1, res);
        sort(res.begin(), res.end());
        return res;
    }
    
};

// { Driver Code Starts.

int main()
{
	int T;
	cin>>T;
	while(T--)
	{
		string S;
		cin>>S;
		Solution ob;
		vector<string> vec = ob.permutation(S);
		for(string s : vec){
		    cout<<s<<" ";
		}
		cout<<endl;
	
	}
	return 0;
}  // } Driver Code Ends

File: 1.                         String Permutations - GFG/1.-string-permutations.py
#User function Template for python3

def toString(List): 
        return ''.join(List) 
    
def permute(a, l, r, res): 
    if l==r: 
        res.append(toString(a))
        return
    
    for i in range(l,r+1): 
        a[l], a[i] = a[i], a[l] 
        permute(a, l+1, r, res) 
        a[l], a[i] = a[i], a[l] 
        
    return 

class Solution:
    
    def permutation(self,s):
        res=[]
        a=list(s)
        n=len(s)
        permute(a, 0, n-1, res)
        res.sort()
        return res

#{ 
#  Driver Code Starts
#Initial Template for Python 3

if __name__=="__main__":
    T=int(input())
    while(T>0):
        s=input()
        ob=Solution()
        ans=ob.permutation(s)
        for i in ans:
            print(i,end=" ")
        print()
        
        T-=1
# } Driver Code Ends